File: infra/darkroom_stack.py
from aws_cdk import (
    CfnOutput,
    Duration,
    Stack,
    aws_iam as iam,
    aws_lambda as lambda_,
)
from constructs import Construct


class DarkroomStack(Stack):
    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        edl_fn = lambda_.Function(
            self,
            "EdlFunction",
            runtime=lambda_.Runtime.PYTHON_3_12,
            handler="handler.handler",
            code=lambda_.Code.from_asset("lambda"),
            timeout=Duration.minutes(5),
            memory_size=512,
            environment={
                "BEDROCK_MODEL_ID": "us.anthropic.claude-sonnet-4-5-20250929-v1:0",
            },
        )

        edl_fn.add_to_role_policy(
            iam.PolicyStatement(
                actions=["bedrock:InvokeModel"],
                resources=[
                    f"arn:aws:bedrock:{self.region}::foundation-model/"
                    "us.anthropic.claude-sonnet-4-5-20250929-v1:0"
                ],
            )
        )

        # IAM-authenticated Function URL — callers need lambda:InvokeFunctionUrl
        fn_url = edl_fn.add_function_url(
            auth_type=lambda_.FunctionUrlAuthType.AWS_IAM,
        )

        CfnOutput(self, "EdlFunctionArn", value=edl_fn.function_arn)
        CfnOutput(self, "EdlFunctionUrl", value=fn_url.url)

File: infra/lib/darkroom-stack.ts
import { CfnOutput, Duration, RemovalPolicy, Stack, StackProps } from 'aws-cdk-lib'
import * as iam from 'aws-cdk-lib/aws-iam'
import * as lambda from 'aws-cdk-lib/aws-lambda'
import * as s3 from 'aws-cdk-lib/aws-s3'
import { Construct } from 'constructs'

export class DarkroomStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props)

    // ── Bedrock EDL Lambda ────────────────────────────────────────────────
    // Thin async orchestration function — only calls boto3 APIs already
    // present in the Lambda runtime, so this stays a plain asset (no Docker
    // bundling required to synth or deploy).
    const edlFn = new lambda.Function(this, 'EdlFunction', {
      runtime: lambda.Runtime.PYTHON_3_12,
      handler: 'handler.handler',
      code: lambda.Code.fromAsset('lambda'),
      timeout: Duration.minutes(5),
      memorySize: 512,
      environment: {
        BEDROCK_MODEL_ID: 'us.anthropic.claude-sonnet-4-5-20250929-v1:0',
      },
    })

    edlFn.addToRolePolicy(
      new iam.PolicyStatement({
        actions: ['bedrock:InvokeModel'],
        resources: [
          `arn:aws:bedrock:${this.region}::foundation-model/` +
            'us.anthropic.claude-sonnet-4-5-20250929-v1:0',
        ],
      }),
    )

    // IAM-authenticated Function URL — callers need lambda:InvokeFunctionUrl
    const edlFnUrl = edlFn.addFunctionUrl({
      authType: lambda.FunctionUrlAuthType.AWS_IAM,
    })

    new CfnOutput(this, 'EdlFunctionArn', { value: edlFn.functionArn })
    new CfnOutput(this, 'EdlFunctionUrl', { value: edlFnUrl.url })

    // ── Amazon Transcribe scratch bucket ────────────────────────────────────
    // Holds audio uploaded for cloud transcription and the resulting
    // transcript JSON when a project's TRANSCRIBE_PROVIDER=aws. Darkroom's
    // backend deletes each object itself right after the job completes —
    // this lifecycle rule is just a backstop against orphaned objects.
    const transcribeBucket = new s3.Bucket(this, 'TranscribeBucket', {
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      lifecycleRules: [{ expiration: Duration.days(1) }],
    })

    new CfnOutput(this, 'TranscribeBucketName', { value: transcribeBucket.bucketName })
  }
}
